src/main/kotlin/build/buf/intellij/resolve/BufRootsProvider.kt

```diff
@@ -14,57 +14,130 @@
+
+        fun getBufCacheFolderV2(env: Map<String, String> = System.getenv()) : VirtualFile? {
+            return LocalFileSystem.getInstance().findFileByNioFile(getBufCacheFolderBase(env))
+                ?.findFileByRelativePath("v2/module")
+        }
+
+        private fun findManifestV2(mod: BufModuleCoordinates, env: Map<String, String> = System.getenv()): Manifest? {
+            val repoCacheFolder = getBufCacheFolderV2(env)
+                ?.findFileByRelativePath("${mod.remote}/${mod.owner}/${mod.repository}") ?: return null
+            return Manifest.fromCommit(repoCacheFolder, mod.commit)
+        }
 
-        fun findModuleCacheFolder(mod: BufModuleCoordinates): VirtualFile? = bufCacheFolder
-            ?.findChild(mod.remote)
-            ?.findChild(mod.owner)
-            ?.findChild(mod.repository)
-            ?.findChild(mod.commit)
+        fun findModuleCacheFolderV1(mod: BufModuleCoordinates, env: Map<String, String> = System.getenv()): VirtualFile? = getBufCacheFolderV1(env)
+            ?.findFileByRelativePath("${mod.remote}/${mod.owner}/${mod.repository}/${mod.commit}")
+
+        fun getOrCreateModuleCacheFolderV2(mod: BufModuleCoordinates, env: Map<String, String> = System.getenv()): VirtualFile? {
+            val bufPath = Paths.get(PathManager.getSystemPath(), "buf")
+            if (!bufPath.isDirectory() && !bufPath.toFile().mkdirs()) {
+                return null
+            }
+            val commitPath = bufPath.resolve("${mod.remote}/${mod.owner}/${mod.repository}/${mod.commit}")
+            if (commitPath.isDirectory()) {
+                return LocalFileSystem.getInstance().refreshAndFindFileByNioFile(commitPath)
+            }
+            val manifest = findManifestV2(mod, env) ?: return null
+            for (path in manifest.getPaths()) {
+                val canonicalPath = manifest.getCanonicalPath(path) ?: return null
+                val modulePath = commitPath.resolve(path)
+                if (!modulePath.parent.isDirectory()) {
+                    Files.createDirectories(modulePath.parent)
+                }
+                Files.copy(Path.of(canonicalPath), modulePath)
+            }
+            return LocalFileSystem.getInstance().refreshAndFindFileByNioFile(commitPath)
+        }
     }
 
     /**
      * Let's index all Buf modules
      * */
     override fun getRootsToWatch(project: Project): Collection<VirtualFile> {
         return listOfNotNull(
-            bufCacheFolder ?: LocalFileSystem.getInstance().findFileByNioFile(bufCacheFolderBase)
+            Paths.get(PathManager.getSystemPath(), "buf").let { LocalFileSystem.getInstance().findFileByNioFile(it) },
+            getBufCacheFolderV2(),
+            getBufCacheFolderV1(),
         )
     }
 
     override fun getAdditionalProjectLibraries(project: Project): Collection<SyntheticLibrary> {
-        return listOf(BufCacheLibrary(bufCacheFolder ?: return emptyList()))
+        if (DumbService.isDumb(project)) {
```

Comment:
This is a state where IntelliJ is in when it rebuilds indexes. This method gets called afterwards with valid index data.